// @builder.io/eslint-plugin-mitosis/dist/rules/__tests__/no-var-name-same-as-state-property.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
exports.__esModule = true;
var eslint_1 = require("eslint");
var no_var_name_same_as_state_property_1 = require("../no-var-name-same-as-state-property");
var opts = {
    filename: 'component.lite.tsx',
    parserOptions: {
        ecmaVersion: 2018,
        sourceType: 'module',
        ecmaFeatures: {
            jsx: true
        }
    }
};
var ruleTester = new eslint_1.RuleTester();
ruleTester.run('no-var-name-same-as-state-property', no_var_name_same_as_state_property_1["default"], {
    valid: [
        __assign(__assign({}, opts), { code: "\n      import { useState } from '@builder.io/mitosis';\n      \n      export default function MyComponent(props) {\n        const state = useState({\n            foo: \"bar\"\n          })\n          \n          const foo_ = bar;\n      \n        return (\n          <div />\n        );\n      }\n      " }),
        __assign(__assign({}, opts), { code: "\n      import { useState } from '@builder.io/mitosis';\n      \n      export default function MyComponent(props) {\n        const state = useState({\n            foo: \"bar\"\n          })\n          \n          const foo = bar;\n      \n        return (\n          <div />\n        );\n      }\n      ", filename: 'file.jsx' }),
    ],
    invalid: [
        __assign(__assign({}, opts), { code: "\n      import { useState } from '@builder.io/mitosis';\n      \n      export default function MyComponent(props) {\n        const state = useState({\n            foo: \"bar\"\n          })\n          \n          const foo = bar;\n      \n        return (\n          <div />\n        );\n      }\n      ", errors: [
                'variables with the same name as a state property will shadow it',
            ] }),
    ]
});
