// @builder.io/mitosis/dist/src/parsers/jsx/helpers.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTypescriptFile = exports.babelDefaultTransform = exports.babelStripTypes = exports.transformAttributeName = exports.HTML_ATTR_FROM_JSX = exports.isImportOrDefaultExport = exports.parseCodeJson = exports.parseCode = exports.uncapitalize = void 0;
const json_1 = require("../../helpers/json");
const typescript_1 = require("../../helpers/typescript");
const babel = __importStar(require("@babel/core"));
const generator_1 = __importDefault(require("@babel/generator"));
const plugin_syntax_typescript_1 = __importDefault(require("@babel/plugin-syntax-typescript"));
const preset_typescript_1 = __importDefault(require("@babel/preset-typescript"));
const typescriptBabelPreset = [preset_typescript_1.default, { isTSX: true, allExtensions: true }];
const { types } = babel;
const uncapitalize = (str) => {
    if (!str) {
        return str;
    }
    return str[0].toLowerCase() + str.slice(1);
};
exports.uncapitalize = uncapitalize;
const parseCode = (node) => {
    const generatorResult = (0, generator_1.default)(node);
    return generatorResult.code;
};
exports.parseCode = parseCode;
const parseCodeJson = (node) => {
    const code = (0, exports.parseCode)(node);
    return (0, json_1.tryParseJson)(code);
};
exports.parseCodeJson = parseCodeJson;
const isImportOrDefaultExport = (node) => types.isExportDefaultDeclaration(node) || types.isImportDeclaration(node);
exports.isImportOrDefaultExport = isImportOrDefaultExport;
exports.HTML_ATTR_FROM_JSX = {
    htmlFor: 'for',
};
const transformAttributeName = (name) => {
    if ((0, typescript_1.objectHasKey)(exports.HTML_ATTR_FROM_JSX, name))
        return exports.HTML_ATTR_FROM_JSX[name];
    return name;
};
exports.transformAttributeName = transformAttributeName;
const babelStripTypes = (code, typescript) => {
    var _a;
    return typescript
        ? (_a = babel.transform(code, {
            configFile: false,
            babelrc: false,
            presets: [typescriptBabelPreset],
        })) === null || _a === void 0 ? void 0 : _a.code
        : code;
};
exports.babelStripTypes = babelStripTypes;
const babelDefaultTransform = (code, visitor) => babel.transform(code, {
    configFile: false,
    babelrc: false,
    comments: false,
    plugins: [[plugin_syntax_typescript_1.default, { isTSX: true }], () => ({ visitor })],
});
exports.babelDefaultTransform = babelDefaultTransform;
const isTypescriptFile = (fileName) => fileName.endsWith('.ts') || fileName.endsWith('.tsx');
exports.isTypescriptFile = isTypescriptFile;
