// @builder.io/mitosis/dist/src/parsers/svelte/typescript/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.collectTypes = exports.parseAttributes = exports.createTagRegex = exports.isTypeScriptComponent = void 0;
const babel = __importStar(require("@babel/core"));
const generator_1 = __importDefault(require("@babel/generator"));
const parser = __importStar(require("@babel/parser"));
const types = __importStar(require("@babel/types"));
const lodash_1 = require("lodash");
function isTypeScriptComponent(string_) {
    const regex = createTagRegex('script', 'gi');
    const match = regex.exec(string_);
    const { lang } = parseAttributes(((match === null || match === void 0 ? void 0 : match.length) && match[1]) || '');
    return lang === 'ts';
}
exports.isTypeScriptComponent = isTypeScriptComponent;
/** Create a tag matching regexp. */
function createTagRegex(tagName, flags) {
    return new RegExp(`/<!--[^]*?-->|<${tagName}(\\s[^]*?)?(?:>([^]*?)<\\/${tagName}>|\\/>)`, flags);
}
exports.createTagRegex = createTagRegex;
/** Transform an attribute string into a key-value object */
function parseAttributes(attributesString) {
    return attributesString
        .split(/\s+/)
        .filter(Boolean)
        .reduce((accumulator, attribute) => {
        const [name, value] = attribute.split('=');
        // istanbul ignore next
        accumulator[name] = value ? value.replace(/["']/g, '') : true;
        return accumulator;
    }, {});
}
exports.parseAttributes = parseAttributes;
function getScriptContent(markup, module) {
    const regex = createTagRegex('script', 'gi');
    let match;
    while ((match = regex.exec(markup)) !== null) {
        const { context } = parseAttributes(match[1] || '');
        if ((context !== 'module' && !module) || (context === 'module' && module)) {
            return match[2];
        }
    }
    return '';
}
function collectTypes(string_, json) {
    var _a;
    const module = getScriptContent(string_, true); // module
    const instance = getScriptContent(string_, false); // instance
    function traverse(script_) {
        const ast = parser.parse(script_, {
            sourceType: 'module',
            plugins: ['typescript'],
        });
        babel.traverse(ast, {
            enter(path) {
                var _a;
                // alias or interface (e.g. type Props = { } or interface Props {} )
                if (types.isTSTypeAliasDeclaration(path.node) ||
                    types.isTSInterfaceDeclaration(path.node)) {
                    json.types = [...((_a = json.types) !== null && _a !== void 0 ? _a : []), (0, generator_1.default)(path.node).code];
                    path.skip();
                }
                else if (types.isTSTypeAnnotation(path.node)) {
                    // add to actual ref
                    const reference = (0, generator_1.default)(path.parent).code;
                    const type = (0, generator_1.default)(path.node.typeAnnotation).code;
                    // add to ref
                    if (Object.prototype.hasOwnProperty.call(json.refs, reference)) {
                        json.refs[reference].typeParameter = type;
                    }
                    // temp add to prop object.
                    // after having finished traversing, we'll create the prop type declaration
                    if (Object.prototype.hasOwnProperty.call(json.props, reference)) {
                        json.props[reference].type = type;
                    }
                }
            },
        });
    }
    traverse(module);
    traverse(instance);
    // add prop type declaration to json.types and set the propsTypeRef
    if ((0, lodash_1.some)(json.props, (property) => !!property.type)) {
        let propertyTypeDeclaration = `type Props = {`;
        propertyTypeDeclaration += Object.keys((0, lodash_1.pickBy)(json.props, (property) => !!property.type))
            .map((key) => {
            return `${key}: ${json.props[key].type};`;
        })
            .join('\n');
        propertyTypeDeclaration += '}';
        json.types = [...((_a = json.types) !== null && _a !== void 0 ? _a : []), propertyTypeDeclaration];
        json.propsTypeRef = 'Props';
    }
}
exports.collectTypes = collectTypes;
