// @builder.io/mitosis/dist/src/targets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.targets = exports.builder = void 0;
const alpine_1 = require("./generators/alpine");
const angular_1 = require("./generators/angular");
const builder_1 = require("./generators/builder");
const html_1 = require("./generators/html");
const liquid_1 = require("./generators/liquid");
const lit_1 = require("./generators/lit");
const marko_1 = require("./generators/marko");
const mitosis_1 = require("./generators/mitosis");
const qwik_1 = require("./generators/qwik");
const react_1 = require("./generators/react");
const react_native_1 = require("./generators/react-native");
const rsc_1 = require("./generators/rsc");
const solid_1 = require("./generators/solid");
const stencil_1 = require("./generators/stencil");
const svelte_1 = require("./generators/svelte");
const swift_1 = require("./generators/swift");
const taro_1 = require("./generators/taro");
const template_1 = require("./generators/template");
const vue_1 = require("./generators/vue");
exports.builder = builder_1.componentToBuilder;
exports.targets = {
    alpine: alpine_1.componentToAlpine,
    angular: angular_1.componentToAngular,
    customElement: html_1.componentToCustomElement,
    html: html_1.componentToHtml,
    mitosis: mitosis_1.componentToMitosis,
    liquid: liquid_1.componentToLiquid,
    react: react_1.componentToReact,
    reactNative: react_native_1.componentToReactNative,
    solid: solid_1.componentToSolid,
    svelte: svelte_1.componentToSvelte,
    swift: swift_1.componentToSwift,
    template: template_1.componentToTemplate,
    webcomponent: html_1.componentToCustomElement,
    vue: vue_1.componentToVue,
    stencil: stencil_1.componentToStencil,
    qwik: qwik_1.componentToQwik,
    marko: marko_1.componentToMarko,
    preact: react_1.componentToPreact,
    lit: lit_1.componentToLit,
    rsc: rsc_1.componentToRsc,
    taro: taro_1.componentToTaro,
};
